server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -490,6 +502,47 @@ private Builder initializeEmpty(IndexMetadata indexMetadata, UnassignedInfo unas
             return this;
         }
 
+        private static List<String> getPreviousNodes(@Nullable IndexRoutingTable previousIndexRoutingTable, int shardId) {
+            if (previousIndexRoutingTable == null) {
+                return null;
+            }
+            final var previousShardRoutingTable = previousIndexRoutingTable.shard(shardId);
+            if (previousShardRoutingTable == null) {
+                return null;
+            }
+            final var primaryNode = previousShardRoutingTable.primaryShard().currentNodeId();
+            if (primaryNode == null) {
+                return null;
+            }
+            final var previousNodes = new ArrayList<String>(previousShardRoutingTable.size());
+            previousNodes.add(primaryNode);
+            for (final var assignedShard : previousShardRoutingTable.assignedShards()) {
```

Comment:
This also includes the target of relocations. I wonder if we should only look at active shards, since anything less will anyway not be considered good enough by the gateway allocator?

The problem I see with this is that if a relocation is ongoing, we risk a copy having a last allocated node id that is much worse than it could be (i.e., a node that only has just started the recovery)?

Reply:
Good point, thanks - see bd12ab97cc6.